[Design] 개인정보 입력 뷰를 구현합니다.

ccdkk opened the pull request:
<!-- 불필요한 항목은 삭제해주세요 -->


## 관련 이슈
- Close #102 
<!-- 관련있는 이슈 번호(#000)을 적어주세요. -->

## 배경

<!-- PR과 관련된 논의 쓰레드, 디자인 가이드, 관련 PR 링크 등을 적어주세요 -->

## 작업 내용
### 1️⃣ 루키의 코드와 & `SelectCollectionView`를 활용하여 구현하였습니다.
루키 코드를 리뷰하면서 보셨을 동일한 형태라 자세한 설명은 생략하겠습니다!!

### 2️⃣ 필수항목이 모두 채워지면 다음 버튼이 Enabled됩니다.
세가지 항목이 모두 true이면 isEnabeld를 true로 바꿔줍니다.
isAvailable은 중복확인 받은 name과 현재 textField의 입력된 내용이 일치하는지 판단하는 함수입니다. 3번 작업내용에서 
자세히 설명하겠습니다.
```Swift
private func checkNextButtonEnabledState() {
        let isAgeSelected = ageSelectCollectionView.isSelected()
        let isGenderSelected = genderSelectCollectionView.isSelected()
        let isAvailableName = userNamingTextField.isAvailableName()
        
        if isAgeSelected && isGenderSelected && isAvailableName {
            self.nextButton.isEnabled = true
        } else {
            self.nextButton.isEnabled = false
        }
    }
```
위 함수는 성별선택, 나이 선택 cell이 눌릴 때마다 & userName의 중복확인을할 때마다 실행합니다.
이를 위해서 공용 컴포넌트인 `TextLimitTextField`, `SelectCollectionView`에 delegate를 추가하였습니다.

### 3️⃣ 중복확인 받은 이름 check 플로우 (feat: @Rookie0031 )
기존의 다음버튼 enabled 함수는 중복체크여부만 확인하다보니, 중복 확인을 받고 난 뒤 이름을 수정하여도 다음버튼이 활성화되어있는 버그가 있었습니다. 이를 해결하기 위해 중복 체크 여부가 아닌 현재 textField에 입력된 값이 중복확인된 이름인지 판단하는 Bool값을 통해 다음버튼의 enabled를 체크합니다.

`availableName`은 optional 값으로, 중복확인이 성공하면 `availableName`에 중복확인 받은 String을 대입합니다.
이후 textField의 변화가 생겨서 기존의 `availableName`값과 달라지게되면 false값을 return합니다.
```Swift
// 선언부
private var availableName: String? = nil
// 관련 함수
    func isAvailableName() -> Bool {
        guard let availableName else { return false }
        return self.textField.text == availableName
    }
```
해당 상태에 맞추어 현재 textField의 text가 중복확인 받은 값과 달라지면 "중복 확인을 체크해주세요" Label을 띄워줍니다.


|수정 전|수정 후|
|---|---|
|![Simulator Screen Recording - iPhone 14 Pro - 2023-02-23 at 16 13 06](https://user-images.githubusercontent.com/91456952/220841444-39b16906-cf6d-4076-883d-45aa668a16c7.gif)|![Simulator Screen Recording - iPhone 14 Pro - 2023-02-23 at 16 14 37](https://user-images.githubusercontent.com/91456952/220841386-29634431-7f5d-4303-993d-951eeb0f7e79.gif)|

<!-- PR 작성 이유와 어떤 변경이 있었는지를 포함합니다. PR에 많은 변경이 있는 경우 추가되는 클래스들의 구조나 동작 등 자세하게 적는 경우도 있습니다 -->

## 리뷰 노트
- delegate로 구현한게 괜찮은지?
- 함수 네이밍, 프로퍼티명이 괜찮은지?
<!-- 구현 시에 고민이었던 점들 혹은 특정 부분에 대한 의도가 있었다면 PR 리뷰의 이해를 돕기 위해 서술합니다. 또한 리뷰어에게 특정 부분에 대한 집중 혹은 코멘트를 요청하는 경우에 작성합니다. -->

## 테스트 방법
`UserInfoInputViewController` 진입점 변경
<!-- PR 리뷰어가 이 PR의 변경사항을 확인할 수 있는 방법을 서술합니다. 의도하는 테스트 결과도 서술하면 더 좋습니다 -->

## 스크린샷
![Simulator Screen Recording - iPhone 14 Pro - 2023-02-22 at 13 59 10](https://user-images.githubusercontent.com/91456952/220526208-5ec359b6-621c-4194-8a9e-91aa7e4486d8.gif)


<!-- 화면 전환이나 인터랙션이 있는 경우 GIF를, 정적인 화면이라면 스크린샷을 첨부합니다. -->
<!-- <img width="" alt="" src=""> -->

<!--## 해당 PR 확정 사항-->

<!-- 해당 PR 리뷰 과정에서 논의된 확정 사항을 develop 브랜치에 머지하기 전 정리합니다. -->

ccdkk committed:
[Add]  UserInfoInputViewController 추가 (#143)

ccdkk committed:
Merge branch 'feature/94-bandinformation-fill' of https://github.com/extreme-rock/GetARock-iOS into feature/143-signup-profile-input-view

ccdkk committed:
[Add] UserInfoInputViewController UI 완성 (#143)

ccdkk committed:
[Add] UIViewController extension 추가 (#143)
- keyboard hide 관련

ccdkk committed:
[Del] keyboard 관련 함수 제거 (#143)

ccdkk committed:
[Del] 충돌방지를 위한 UIViewController Extension 삭제 (#102)

ccdkk committed:
[Style] 진입점 변경 print문 삭제 (#102)

ccdkk committed:
Merge branch 'develop' of https://github.com/extreme-rock/GetARock-iOS into feature/102-signup-profile-input-view

ccdkk committed:
[Add] keyboard hide 함수 실행 (#102)

ccdkk committed:
[Feat] 다음버튼 활성화를 위한 델리게이트 패턴 추가 (#102)

ccdkk committed:
[Feat] 키보드 관련 함수 추가 (#102)

ccdkk committed:
[Design] 마진 변경 (#102)

GODNOEL commented on GetARock/GetARock/Global/Supports/SceneDelegate.swift:
```diff
@@ -19,7 +19,7 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         
         let window = UIWindow(windowScene: windowScene)
-        window.rootViewController = PositionSelectViewController()
+        window.rootViewController = UserInfoInputViewController()
```
p5;
나중에 merge할때 랜딩페이지로 원복해주세요!

ccdkk replied:
[74f99cb](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/74f99cb5f0ae4dbc07e06efcdf6c603da55eacd3) 반영완료!

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+    private let userNamingGuideSubLabel = BasicLabel(
+        contentText: "* 공백없이 20자 이하, 기호는 _만 입력 가능합니다.",
+        fontStyle: .content,
+        textColorInfo: .gray02)
```
p4;
fontStyle `caption`으로 변경해주실 수 있나요?
안내정보들은 제일 작은 `caption` 폰트를 사용하고 있습니다!

<img width="600" alt="image" src="https://user-images.githubusercontent.com/103012104/220802326-3b8d7cdd-f909-470f-96b5-dbaa13ffc893.png">

ccdkk replied:
[3a6ac71](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/3a6ac71ed7759adbd11335016d43d548afc86de5)반영완료!

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+        duplicationCheckType: .userName,
+        textExpressionCheck: true)
+    
+    private lazy var textFieldStackView: UIStackView = {
```
p5;
`textFieldStackView` 를 좀 더 직관적으로 `userNameInputStackView`로 변경하는건 어떤가요?
지금은 어떤 뷰인지 바로 알아차리기 함든것같아서요..!

ccdkk replied:
너무 스택뷰 네이밍만 통일감이 없었네요! [3a6ac71](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/3a6ac71ed7759adbd11335016d43d548afc86de5)반영완료!

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+        $0.spacing = 10
+        return $0
+    }(UIStackView(arrangedSubviews: [userIntroGuideTitleLabel,
+                                     userIntroTextView]))
```
p5;
textViewStackView 를 좀 더 직관적으로 `userIntroInputStackView`로 변경하는건 어떠신가요?

ccdkk replied:
[3a6ac71](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/3a6ac71ed7759adbd11335016d43d548afc86de5)반영완료!

Rookie0031 commented on GetARock/GetARock/Global/Supports/SceneDelegate.swift:
```diff
@@ -19,7 +19,7 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         
         let window = UIWindow(windowScene: windowScene)
-        window.rootViewController = PositionSelectViewController()
+        window.rootViewController = UserInfoInputViewController()
```
p5;

시작점이 바뀐채로 커밋되어있습니다!

ccdkk replied:
[74f99cb](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/74f99cb5f0ae4dbc07e06efcdf6c603da55eacd3) 반영완료!

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+        contentText: "* 본인계정이 아닌 계정 등록 시 책임은 본인에게 있습니다.",
+        fontStyle: .content,
+        textColorInfo: .gray02
+    )
```
p4;
fontStyle caption으로 변경해주실 수 있나요?
안내정보들은 제일 작은 caption 폰트를 사용하고 있습니다!

ccdkk replied:
[3a6ac71](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/3a6ac71ed7759adbd11335016d43d548afc86de5)반영완료!

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+
+    private let instagramTextField = SNSBoxView(type: .instagram, placeholder: "사용자 계정")
+
+    private let soundCloudTextField = SNSBoxView(type: .soundCloud, placeholder: "사용자 계정")
```
p5;
혹시 의도하신게 아니라면 프로퍼티 간의 개행 없애주실 수 있나요?

ccdkk replied:
넵! [3a6ac71](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/3a6ac71ed7759adbd11335016d43d548afc86de5)반영완료!

Rookie0031 commented on GetARock/GetARock/Global/UIComponent/SelectCollectionView.swift:
```diff
@@ -74,6 +80,13 @@ final class SelectCollectionView: UIView {
+    
+    func isSelected() -> Bool {
+        guard let selectedIndex = collectionView.indexPathsForSelectedItems else { return false}
+        let isSelected = !(selectedIndex.isEmpty)
```
p5;

가독성을 위해서 따로 중간에 변수에 저장하신건가요??

ccdkk replied:
넵 맞습니다. !selectedIndex.isEmpty 가 isSelected라고 바로 연상이 안된다고 생각하여 따로 지정했습니다.

GODNOEL commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+                                     genderInputStackView,
+                                     textViewStackView,
+                                     snsInformationStackView,
+                                     informationFillCompleteButton]))
```
p;
모두 스택뷰에 넣어버린거 너무 아름답네요🥹

![image](https://user-images.githubusercontent.com/103012104/220803876-749ddcbd-11ad-432f-abd8-f2a01a1a7115.png)

ccdkk replied:
갓 루키.. UIStackView로 contentView 하는건 루키 코드에서 첨봤는데 layout함수가 깔끔해져서 참 좋네요

Rookie0031 commented on GetARock/GetARock/Screens/ModifyMyPage/ModifyUserProfileViewController.swift:
```diff
@@ -8,8 +8,207 @@
+                                     soundCloudTextField]))
+
+    private let informationFillCompleteButton: BottomButton = {
+        //TODO: 밴드 정보 POST action 추가 필요
```
p5;

여기 개인정보 입력뷰에 맞게 주석을 바꿔주면 좋을거같습니다!

ccdkk replied:
[a7a99d1](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/a7a99d15db5fb148390d72f986808def4be3287d)에서 반영 완료!

Rookie0031 commented on GetARock/GetARock/Screens/SignUp/UserInfoInputViewController.swift:
```diff
@@ -0,0 +1,259 @@
+    
+    private let ageTitleLabel = InformationGuideLabel(guideText: "연령대", type: .required)
+    
+//    private let ageSelectCollectionView = SelectCollectionView(
```
p5;

불필요한 주석이라면 삭제 부탁드립니다~!

ccdkk replied:
[a7a99d1](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/a7a99d15db5fb148390d72f986808def4be3287d)에서 반영 완료!

Rookie0031 commented on GetARock/GetARock/Screens/SignUp/UserInfoInputViewController.swift:
```diff
@@ -0,0 +1,259 @@
+    }
+    
+    private func checkNextButtonEnabledState() {
+        let isAgeSelected = ageSelectCollectionView.isSelected()
```
p5;

추후에는 입력하지 않은 정보는 alert로 알려주는 코드가 추가되면 좋을거같네요! 주석으로 남겨두면 어떨까요??

ccdkk replied:
오 좋은데요. 내일 디자이너와 얘기하여 필요하면 바로 수정하겠습니다~!

ccdkk replied:
2/24일 회의에서 안하는 것으로 땅땅

Rookie0031 replied:
토의 결과로 안해도 될거같습니다 ㅋㅋㅋ 감사합니다!

GODNOEL reviewed (commented):
## 고생하셨습니다 데이크!
항목이 많아서 제작하시는데 힘드셨을것같은데 뚝딱하셨네여~!
디자인관련 리뷰달았는데 참고해서 Modify페이지에도 적용해주시면 너무너무 감사하겠습니다!
나머지는 도착해서 좀 천천히 보도록 하겠습니다..!

### 🚨 AutoLayout 깨짐
여기도 이전과 동일하게 CollectView와 Cell 둘 다 width 값이 들어가서 
콘솔창에 AutoLayout 깨진다고 뜨더라구요? 
`UIComponent` > `SelectCollectionViewCell` 으로 가셔서 아래 라인 지워주세요!

<img width="1130" alt="image" src="https://user-images.githubusercontent.com/103012104/220804847-f49e7423-638b-4d76-bac5-c04289fb4e02.png">

<br/>

### 🎨 collectionVIew Cell Color
성별 collectionVIew backgroundColor이 다른거 같은데 `dark02` 인지 확인해주실 수 있나여??

<img width="300" alt="image" src="https://user-images.githubusercontent.com/103012104/220805124-6dbb4d68-b6e6-49c9-b313-30cf8bc557eb.png">

Rookie0031 commented on GetARock/GetARock/Screens/SignUp/UserInfoInputViewController.swift:
```diff
@@ -0,0 +1,259 @@
+        let isGenderSelected = genderSelectCollectionView.isSelected()
+        let isDuplicated = userNamingTextField.isNameDuplicated()
+        
+        if isAgeSelected && isGenderSelected && isDuplicated {
```
p1;

여기서 텍스트필드가 isEmpty가 아닌 조건이 추가되어야할 거 같습니다!
중복확인 한 다음, 텍스트필드 글자를 다 지워도 여전히 버튼이 활성화되어있어서 닉네임이 비어있는 채로 생성이 되어버릴 수 있네요.. 가능성은 낮지만 예방하면 좋을거같습니다!

ccdkk replied:
수정사항 pr내용에 추가하였습니다.

Rookie0031 replied:
굿입니다 감사합니다!!

Rookie0031 commented on GetARock/GetARock/Global/UIComponent/SelectCollectionView.swift:
```diff
@@ -99,3 +112,12 @@ extension SelectCollectionView: UICollectionViewDataSource {
 }
+
+extension SelectCollectionView: UICollectionViewDelegate {
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
```
p3;

이 함수도 결국엔 선택 여부를 판별하는 함수라서 역할을 명시할 수 있는 다른 이름은 어떨까요?? checkIsSelected 같은 ??
또한 파라미터로 collection view, indexpath를 전달하는 이유가 궁금하네요? 누를 때 마다 선택여부 체크가 목적이라면 인자를 전달하지않아도 되잖아요?? 예를 들어 이렇게?


    extension SelectCollectionView: UICollectionViewDelegate {
        func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
            self.delegate?.checkIsSelected()
        }
    }

ccdkk replied:
델리게이트 함수명을 기본 API처럼 네이밍하고 싶어서 해당 함수라 불리는 상황을 네이밍에 담으려했습니다.
루키말씀처럼 collectionView나 indexPath는 사용될 것 같지 않아서 다음과 collectionViewDidSelected로 수정했는데 어떤가요?
[1437332](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/1437332567a6128036c1e3da03367060cfe39cac)

Rookie0031 replied:
좋습니다!!

Rookie0031 reviewed (changes requested):
## 고생하셨습니다!! 👍

필수 정보 입력, 선택 여부를 isSelected, is ~~ 같은 이름으로 통일한뒤 최종 체크해서 버튼을 활성화시키는 구조가 좋네요 가독성도 좋은거같아요! 저도 추후 PR에 사용하도록 하겟습니다 굿! 👍

감사합니다 👏

ccdkk committed:
[Del] 보라색 layout 이슈 해결을 위해 width layout 삭제 (#102)

ccdkk committed:
[Design] backgroundColor 사용자화 (#102)

ccdkk committed:
[Style] font 변경 & 네이밍 변경 (#102)

ccdkk committed:
[Style] 진입점 변경 (#102)

ccdkk commented:
@GODNOEL 
> 🚨 AutoLayout 깨짐

항상 감사드립니다.  [169ba50](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/169ba504cc4e0d54ad4237be6ade6a5f5e7eb077)에서 layout 중복 해결했습니다!

> 🎨 collectionVIew Cell Color

요거는 공통 컴포넌트 구현할 때 두 종류에 대하여 backgroundColor가 고정인줄 알고 구현했는데 각 VC에 맞춰서 다르게 사용되는군요!
View Init 시점에 color를 입력하게 하여 사용자화 하였습니다. [a947501](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/a94750103aaf81c92b8737f470c3061897161b88) 확인 부탁드려요!

ccdkk committed:
[Add] 주석 변경 & 주석 추가 & 필요없는 내용 삭제 (#102)

ccdkk committed:
[Feat] nextButton enabled 로직 수정 (#102)
- 중복확인된 string을 저장하고 변화가 생기면 false로

Rookie0031 committed:
[Feat] 중복 체크 확인 이후 텍스트 변경시, 중복체크 재요구 기능 추가 (#102)

Rookie0031 committed:
[Fix] 텍스트 필드 중복 체크시 Request baseURL 변경 (#102)

https 프로토콜 적용에 따라 서버 주소 변경
[참고]
https://getarock.slack.com/archives/C04H26ZRQ77/p1677119864525019

ccdkk committed:
[Del] 프린트문 삭제 (#102)

ccdkk committed:
[Style] 델리게이트명 변경 (#102)

ccdkk committed:
Merge branch 'develop' of https://github.com/extreme-rock/GetARock-iOS into feature/102-signup-profile-input-view

GODNOEL commented:
> 🎨 collectionVIew Cell Color
> 
> 요거는 공통 컴포넌트 구현할 때 두 종류에 대하여 backgroundColor가 고정인줄 알고 구현했는데 각 VC에 맞춰서 다르게 사용되는군요! View Init 시점에 color를 입력하게 하여 사용자화 하였습니다. [a947501](https://github.com/extreme-rock/GetARock-iOS/pull/135/commits/a94750103aaf81c92b8737f470c3061897161b88) 확인 부탁드려요!

어엇 이 셀들은 다 `dark02`로 디자인 했던것같은데, 색이 분기처리되서 사용된 경우가 있었나요?

ccdkk commented:
@GODNOEL 
가입하지 않은 유저를 밴드에 추가할 때 .dark03 으로 사용되는 것 같아요

Rookie0031 reviewed (approved):
꼼꼼한 체크 감사합니다~~ 고생하셧습니다!!

ccdkk committed:
Merge branch 'develop' of https://github.com/extreme-rock/GetARock-iOS into feature/102-signup-profile-input-view
